### DISCUSSION TITLE
solving mirrored font from c_to_python_font on a st7789 2.8 restouch screen

### DISCUSSION THREAD

**Discussion by TLmv** [micropython/micropython#12198] [RP2040 / Pico]
i was trying to get a color image on my pico, but it appeared to be heavy on the pico and on me ;)
and looking at c_to_python_font.py i was hoping to turn my initial idea into some pixel art.

it's a regular pico, running 1.9.1 with a 2.8 inch 240x320 touchres screen that uses the st7789_4bit driver in portrait mode (VER) with nanogui.

i used an online tool to create the bytes from some font art,  ran c_to_python_font and tried the file.
first off, i don't have this problem if i use font_to_py with a TTF file, it only appears to be with B/W images converted to bytes.
the result was messed up, tried vertical sequence of bytecode, but that made it worse.

to check this i created a font with the word [TEST] repeated on it.
all my checks showed it was just the word test.
after using Cwriter to put it on the display it turns out quite different.

it is not really mirrored, but vertical lines mirrored segments. this is best shown on the bottomline, where test reads from left to right, but all letters are morrored.

there should be a question in here about if it is possible to solve this :)

**Comment by peterhinch** [comment]
Firstly does nanogui work with its supplied fonts? If it does, the fault must lie in `c_to_python_font.py`. I'm not familiar with this, have you a webref? It's most unlikely to produce fonts that can be used directly with nanogui.

Fonts produced by [font-to-py](https://github.com/peterhinch/micropython-font-to-py) are in a specific format designed to reduce RAM usage in MicroPython. If you want to produce fancy fonts, I recommend finding a utility which outputs those fonts in OTF or TTF format, then using font-to-py to convert them.

**Comment by TLmv** [comment]
nanogui works with it's supplied fonts and with some fonts i made with use of font-to-py. 
will try your suggestion to first create a ttf from the images and try font_to_py with that and let you know.

i used: https://javl.github.io/image2cpp/ with setting black as background, binary dithering and 1 bit horizontal

**Comment by TLmv** [comment]
c_to_python_font is part of the font-to-py download.  it has a readme instruction in the icon_fonts folder.

if i set image2cpp to swap bits in bytes it looks better, now only a third is presented twice where the first is horizontally mirrored, but it does not finish the bytecode lettertype, like with the ones that did not swap bits in bytes.

thank you for your feedback, font_to_py solved it with some minor atlerations

**Reply by peterhinch** [reply]
> c_to_python_font is part of the font-to-py download

OK, now I see what you were doing. That utility is intended to be used with the Linux `bitmap` application which produces C code as output. Given a set of output files, `c_to_python_font` produces a Python font file. The utility is most unlikely to work with any application other than `bitmap`.

**Comment by TLmv** [comment]
It took me some time, 
but i wanted more from the intended use.

although i had some trouble with the frame buffer size running out when creating a 240x360 default character, even with bdcolor=false. 
i did manage to use it to show a 236x316 'character' and some smaller ones.
alas i couldn't leave it in the background.

this combined with a 3d printed housing made my new PULSE intervalometer.

So thank you for all the help in my understanding the capabilities and limitations.
i'm now busy with another project that has tossed up some hardware related issues that i haven't all solved yet, 
but it also uses nanogui and i'm sure i will make some new lettertype to go with it. 

If you like to see what the result is of my use of nanogui, i made a small 'promotional' film of the finished intervalometer tool and that includes filmed footage from the display as it looks now, so color representation is not ideal:
 
https://www.youtube.com/watch?v=u0uCjxv_uFE

**Reply by peterhinch** [reply]
Nice use of nanogui. 

Your URL format is wrong: clicking on it gives a 404 and you have to copy-paste the text to view the video.
